### ISSUE TITLE
[BUG] Categorify problems when using num_buckets

### ISSUE THREAD

**Issue by albert17** [NVIDIA/NVTabular#543]
@rnyak @benfred 

I am updating the Criteo-HugeCTR Notebook example to use the new API: https://github.com/NVIDIA/NVTabular/blob/764b8960f8bb66ec98d785497f62c50c1457efc8/examples/hugectr/criteo-hugectr.ipynb

Previously, I was `proc.add_cat_preprocess([ops.Categorify(out_path=stats_path), ops.LambdaOp(op_name="MOD10M", f=lambda col: col % num_buckets)])` for the categoricals. 

1. If try to use LambdaOp with the new API: `cat_features = CATEGORICAL_COLUMNS >> Categorify(out_path=stats_path) >> LambdaOp(f=lambda col: col % num_buckets)` I get an empty dict when I try to get the embeddings sizes `get_embedding_sizes(workflow)`.

2. If I try to use `num_buckets` within Categorify `cat_features = CATEGORICAL_COLUMNS >> Categorify(out_path=stats_path, num_buckets=num_buckets)` I get an error due to num_buckets being an int (it seems to be a bug not considering the int case)
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-18-671dd4884770> in <module>
----> 1 embeddings = get_embedding_sizes(workflow)
      2 print(embeddings)
/NVTA/nvtabular/ops/categorify.py in get_embedding_sizes(workflow)
    387         current = queue.pop()
    388         if current.op and hasattr(current.op, "get_embedding_sizes"):
--> 389             output.update(current.op.get_embedding_sizes(current.columns))
    390 
    391             if hasattr(current.op, "get_multihot_columns"):
/NVTA/nvtabular/ops/categorify.py in get_embedding_sizes(self, columns)
    357 
    358     def get_embedding_sizes(self, columns):
--> 359         return _get_embeddings_dask(self.categories, columns, self.num_buckets, self.freq_threshold)
    360 
    361     def get_multihot_columns(self):
/NVTA/nvtabular/ops/categorify.py in _get_embeddings_dask(paths, cat_names, buckets, freq_limit)
    414         path = paths.get(col)
    415         num_rows = cudf.io.read_parquet_metadata(path)[0] if path else 0
--> 416         if buckets and col in buckets and freq_limit[col] > 1:
    417             num_rows += buckets.get(col, 0)
    418         if buckets and col in buckets and not freq_limit[col]:
TypeError: argument of type 'int' is not iterable
```

3. As a workaround for 2, I can build a dictionary specifying the num_buckets for every column, this works, but the embedding sizes are not what I was expecting. 
```
{'C1': (10000000, 512),
 'C10': (10000000, 512),
 'C11': (10000000, 512),
 'C12': (10000000, 512),
 'C13': (10000000, 512),
 'C14': (10000000, 512),
 'C15': (10000000, 512),
 'C16': (10000000, 512),
 'C17': (10000000, 512),
 'C18': (10000000, 512),
 'C19': (10000000, 512),
 'C2': (10000000, 512),
 'C20': (10000000, 512),
 'C21': (10000000, 512),
 'C22': (10000000, 512),
 'C23': (10000000, 512),
 'C24': (10000000, 512),
 'C25': (10000000, 512),
 'C26': (10000000, 512),
 'C3': (10000000, 512),
 'C4': (10000000, 512),
 'C5': (10000000, 512),
 'C6': (10000000, 512),
 'C7': (10000000, 512),
 'C8': (10000000, 512),
 'C9': (10000000, 512)}
```
It seems that `num_buckets `being 10M is replacing the cardinality for every column. However, I was expecting that only the columns its cardinality is greater than `num_buckets `would be limited, and for the rest it would be the same. This is the behaviors I am getting with the old API and Categorify+LamdaOp.

**Comment by rnyak** [comment]
@benfred @albert17 for viz.
Ben, I explained Alberto that it is expected that he is going to see ` num_buckets being 10M is replacing the cardinality for every column.` He expects that for the low cardinalities, after hashing, the embedding table shape would be `(original n_cats, emb_dim)` but I think it will be `(num_buckets, emd_dim)`. so for example, say he is hashing column `C6` which has only 6 unique features, so cardinality =6. he is hashing it with hash_bucket = 10000000. He expects, the final embedding table shape for `C6` would be `(6, emb_dim)` but he gets `(10000000, embd_dim)`. But this is expected. I dont think there is something wrong here.

What's your opinion? please correct me if I miss something.

**Comment by rnyak** [comment]
@albert17 you can also use `HashBucket` op if you are only doing simple hashing. you can still return emb_size for embedding table, it s now implemented. see the link.

https://github.com/NVIDIA/NVTabular/blob/main/nvtabular/ops/hash_bucket.py#L96-L102

**Comment by albert17** [comment]
We tried HashBucket in the past, but we came up to the conclusion that it was better to use LambdaOp. See discussion here: https://github.com/NVIDIA/NVTabular/issues/402

**Comment by rnyak** [comment]
> We tried HashBucket in the past, but we came up to the conclusion that it was better to use LambdaOp. See discussion here: #402

@albert17 I check the issue you linked. Are you using `LambdaOp(lambda col, gdf: gdf[col].hash_values() % bucket_size)` or `LambdaOp(f=lambda col: col % num_buckets)`. These are two different things. The first one will generate `hash_values` first, end then do modulo operator. You will end up a value between [0, 10M) for each categorical feature if you do the first one. Like in this example: 

```
gdf = cudf.DataFrame({'platform':[3, 3, 12, 12, 1, 2, 0, 1]})

bucket_size = 10000000
f = lambda gdf: gdf['platform'].hash_values() % bucket_size
gdf['hashed_platform'] = f(gdf)
print(gdf)
   platform  hashed_platform
0         3          8043756
1         3          8043756
2        12          6190381
3        12          6190381
4         1          2460029
5         2          3430573
6         0          9140149
7         1          2460029
```
so using `LambdaOp(lambda col, gdf: gdf[col].hash_values() % bucket_size)` and using `Hash_Bucket(num_buckets=10M)` is same.

**Comment by albert17** [comment]
I was doing `ops.LambdaOp(op_name="MOD10M", f=lambda col: col % num_buckets)` 

From the past disscusion I understood we wanted to avoid `hash_values()`, we just want to apply the modulo 10M

**Comment by albert17** [comment]
@benfred @rnyak Does it makes sense that when applying `cat_features = CATEGORICAL_COLUMNS >> Categorify(out_path=stats_path) >> LambdaOp(f=lambda col: col % num_buckets)` when I call `get_embedding_sizes` I get an empty dict?

**Comment by albert17** [comment]
Following your example @rnyak. It seems like there is some problem with LambdaOp
```
>>> import nvtabular as nvt
>>> import cudf
>>> gdf = cudf.DataFrame({'platform':[3, 3, 12, 12, 1, 2, 0, 1]})
>>> gdf
   platform
0         3
1         3
2        12
3        12
4         1
5         2
6         0
7         1
>>> dataset = nvt.Dataset(gdf)

# Categorify + LamdaOp
>>> cat_features = "platform" >> nvt.ops.Categorify() >> nvt.ops.LambdaOp(f=lambda col: col % 10)
>>> workflow = nvt.Workflow(cat_features)
>>> wrokflow.fit(dataset)
>>> workflow.transform(dataset)
<nvtabular.io.dataset.Dataset object at 0x7fe2af5ba9d0>
>>> nvt.ops.get_embedding_sizes(workflow)
{}

# Only Categorify
>>> cat_features = "platform" >> nvt.ops.Categorify()
>>> workflow = nvt.Workflow(cat_features)
>>> workflow.fit(dataset)
>>> workflow.transform(dataset)
<nvtabular.io.dataset.Dataset object at 0x7fe2af5bab50>
>>> nvt.ops.get_embedding_sizes(workflow)
{'platform': (6, 16)}

# Categorify + HashBucket
>>> cat_features = "platform" >> nvt.ops.Categorify() >> nvt.ops.HashBucket(10)
>>> workflow = nvt.Workflow(cat_features)
>>> workflow.fit(dataset)
>>> workflow.transform(dataset)
<nvtabular.io.dataset.Dataset object at 0x7fe2c020a610>
>>> nvt.ops.get_embedding_sizes(workflow)
{'platform': (10, 16)}
```

Using just Categorify or Categorify + HashBucket I get the embedding table sizes; however, if I use Categorify + LamdaOp I get an empty dict.

**Comment by albert17** [comment]
I have tried using HashBucket, and I am getting the same result than with Categorify.
```
{'C1': (10000000, 512),
 'C10': (10000000, 512),
 'C11': (10000000, 512),
 'C12': (10000000, 512),
 'C13': (10000000, 512),
 'C14': (10000000, 512),
 'C15': (10000000, 512),
 'C16': (10000000, 512),
 'C17': (10000000, 512),
 'C18': (10000000, 512),
 'C19': (10000000, 512),
 'C2': (10000000, 512),
 'C20': (10000000, 512),
 'C21': (10000000, 512),
 'C22': (10000000, 512),
 'C23': (10000000, 512),
 'C24': (10000000, 512),
 'C25': (10000000, 512),
 'C26': (10000000, 512),
 'C3': (10000000, 512),
 'C4': (10000000, 512),
 'C5': (10000000, 512),
 'C6': (10000000, 512),
 'C7': (10000000, 512),
 'C8': (10000000, 512),
 'C9': (10000000, 512)}
```

**Comment by rnyak** [comment]
> I have tried using HashBucket, and I am getting the same result than with Categorify.
> 
> ```
> {'C1': (10000000, 512),
>  'C10': (10000000, 512),
>  'C11': (10000000, 512),
>  'C12': (10000000, 512),
>  'C13': (10000000, 512),
>  'C14': (10000000, 512),
>  'C15': (10000000, 512),
>  'C16': (10000000, 512),
>  'C17': (10000000, 512),
>  'C18': (10000000, 512),
>  'C19': (10000000, 512),
>  'C2': (10000000, 512),
>  'C20': (10000000, 512),
>  'C21': (10000000, 512),
>  'C22': (10000000, 512),
>  'C23': (10000000, 512),
>  'C24': (10000000, 512),
>  'C25': (10000000, 512),
>  'C26': (10000000, 512),
>  'C3': (10000000, 512),
>  'C4': (10000000, 512),
>  'C5': (10000000, 512),
>  'C6': (10000000, 512),
>  'C7': (10000000, 512),
>  'C8': (10000000, 512),
>  'C9': (10000000, 512)}
> ```

yes, this is expected, and there is no problem here.

**Comment by rnyak** [comment]
> Following your example @rnyak. It seems like there is some problem with LambdaOp
> 
> ```
> >>> import nvtabular as nvt
> >>> import cudf
> >>> gdf = cudf.DataFrame({'platform':[3, 3, 12, 12, 1, 2, 0, 1]})
> >>> gdf
>    platform
> 0         3
> 1         3
> 2        12
> 3        12
> 4         1
> 5         2
> 6         0
> 7         1
> >>> dataset = nvt.Dataset(gdf)
> 
> # Categorify + LamdaOp
> >>> cat_features = "platform" >> nvt.ops.Categorify() >> nvt.ops.LambdaOp(f=lambda col: col % 10)
> >>> workflow = nvt.Workflow(cat_features)
> >>> wrokflow.fit(dataset)
> >>> workflow.transform(dataset)
> <nvtabular.io.dataset.Dataset object at 0x7fe2af5ba9d0>
> >>> nvt.ops.get_embedding_sizes(workflow)
> {}
> 
> # Only Categorify
> >>> cat_features = "platform" >> nvt.ops.Categorify()
> >>> workflow = nvt.Workflow(cat_features)
> >>> workflow.fit(dataset)
> >>> workflow.transform(dataset)
> <nvtabular.io.dataset.Dataset object at 0x7fe2af5bab50>
> >>> nvt.ops.get_embedding_sizes(workflow)
> {'platform': (6, 16)}
> 
> # Categorify + HashBucket
> >>> cat_features = "platform" >> nvt.ops.Categorify() >> nvt.ops.HashBucket(10)
> >>> workflow = nvt.Workflow(cat_features)
> >>> workflow.fit(dataset)
> >>> workflow.transform(dataset)
> <nvtabular.io.dataset.Dataset object at 0x7fe2c020a610>
> >>> nvt.ops.get_embedding_sizes(workflow)
> {'platform': (10, 16)}
> ```
> 
> Using just Categorify or Categorify + HashBucket I get the embedding table sizes; however, if I use Categorify + LamdaOp I get an empty dict.

My understanding, applying `Categorify + Lambda` op would not change  the embedding table shape the one that you'd get from only `Categorify` op. Because, `Lambda` op does not have a link with `get_emb_size` function and it does not collect stats to do so. However, you should still get an embedding table from `nvt.ops.get_embedding_sizes(workflow)`, because you use `Categorify` before `Lambda` op. It should print out an embedding table.

**Comment by albert17** [comment]
This is how I was using it before
```
num_buckets = 10000000
proc.add_cat_preprocess([ops.Categorify(out_path=stats_path), ops.LambdaOp(op_name="MOD10M", f=lambda col: col % num_buckets)])
```

And I was getting the embedding sizes like this
```
embeddings = [c[0] for c in ops.get_embedding_sizes(workflow).values()]
embeddings = np.clip(a=embeddings, a_min=None, a_max=num_buckets).tolist()
print(embeddings)
```

And the result was looking as expected
```

[10000000, 10000000, 3014529, 400781, 11, 2209, 11869, 148, 4, 977, 15, 38713, 10000000, 10000000, 10000000, 584616, 12883, 109, 37, 17177, 7425, 20266, 4, 7085, 1535, 64]
```

**Comment by benfred** [comment]
@albert17 with the new api changes you can do something like:

```python
categorify_op = nvtabular.ops.Categorify()
cat_features = CAT_COLUMNS >> categorify_op >> (lambda col: col % num_buckets)

workflow = nvtabular.Workflow(cat_features + other_features)
workflow.fit(dataset)

embeddings = {col: (min(count, num_buckets), dim) for col, (count, dim) in 
                            categorify_op.get_embedding_sizes(CAT_COLUMNS).items()}
```

to achieve the same result. Note that the global 'get_embedding_sizes' from the workflow isn't working anymore because of the lambda op after the categorify - https://github.com/NVIDIA/NVTabular/blob/4606b0c7d8ce81f79175db432ddc429fa4fa0c87/nvtabular/ops/categorify.py#L395-L397 -

**Comment by benfred** [comment]
We should also implement fixed sized embedding sizes in #194

**Comment by benfred** [comment]
fixed by https://github.com/NVIDIA/NVTabular/pull/544
